contracts/staking/contracts/src/fees/MixinExchangeFees.sol

```diff
@@ -0,0 +1,245 @@
+        if (_feesCollectedThisEpoch == 0) {
+            activePoolsThisEpoch.push(poolId);
+        }
+    }
+
+    /// @dev Pays the rebates for to market making pool that was active this epoch,
+    /// then updates the epoch and other time-based periods via the scheduler (see MixinScheduler).
+    /// This is intentionally permissionless, and may be called by anyone.
+    function finalizeFees()
+        external
+    {
+        // distribute fees to market maker pools as a reward
+        (uint256 totalActivePools,
+        uint256 totalFeesCollected,
+        uint256 totalWeightedStake,
+        uint256 totalRewardsPaid,
+        uint256 initialContractBalance,
+        uint256 finalContractBalance) = _distributeFeesAmongMakerPools();
+        emit RewardsPaid(
+            totalActivePools,
+            totalFeesCollected,
+            totalWeightedStake,
+            totalRewardsPaid,
+            initialContractBalance,
+            finalContractBalance
+        );
+
+        _goToNextEpoch();
+    }
+
+    /// @dev Returns the total amount of fees collected thus far, in the current epoch.
+    /// @return Amount of fees.
+    function getTotalProtocolFeesThisEpoch()
+        public
+        view
+        returns (uint256)
+    {
+        return address(this).balance;
+    }
+
+    /// @dev Returns the amount of fees attributed to the input pool.
+    /// @param poolId Pool Id to query.
+    /// @return Amount of fees.
+    function getProtocolFeesThisEpochByPool(bytes32 poolId)
+        public
+        view
+        returns (uint256)
+    {
+        return protocolFeesThisEpochByPool[poolId];
+    }
+
+    /// @dev Pays rewards to market making pools that were active this epoch.
+    /// Each pool receives a portion of the fees generated this epoch (see LibFeeMath) that is
+    /// proportional to (i) the fee volume attributed to their pool over the epoch, and 
+    /// (ii) the amount of stake provided by the maker and their delegators. Rebates are paid
+    /// into the Reward Vault (see MixinStakingPoolRewardVault) where they can be withdraw by makers and
+    /// the members of their pool. There will be a small amount of ETH leftover in this contract
+    /// after paying out the rebates; at present, this rolls over into the next epoch. Eventually,
+    /// we plan to deposit this leftover into a DAO managed by the 0x community.
+    /// @return totalActivePools Total active pools this epoch.
+    /// @return totalFeesCollected Total fees collected this epoch, across all active pools.
+    /// @return totalWeightedStake Total weighted stake attributed to each pool. Delegated stake is weighted less.
+    /// @return totalRewardsPaid Total rewards paid out across all active pools.
+    /// @return initialContractBalance Balance of this contract before paying rewards.
+    /// @return finalContractBalance Balance of this contract after paying rewards.
+    function _distributeFeesAmongMakerPools()
```

Comment:
We should add a TODO that this function needs a way to be executed among multiple blocks (or reduced to ~O(1) complexity). This will be a hard requirement, but can be done in another PR.